File: Dynamic Programming/knapSack.java
import java.util.*;

class knapSack {
    static void display(int[][] array, int noOfItems, int totalCapacity) {
        System.out.print("\nThe knapsack table using dynamic programming is as follows:\n");
        for (int i = 0; i <= noOfItems; i++) {
            for (int w = 0; w <= totalCapacity; w++) {
                System.out.print(array[i][w] + "\t");
            }
            System.out.println();
        }
    }
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        System.out.print("\nEnter the number of items in the knapsack problem: ");
        int noOfItems = in.nextInt();
        System.out.print("Enter the total capacity of the knapsack problem: ");
        int totalCapacity = in.nextInt();
        
        int[] weightOfItems = new int[noOfItems];
        int[] profitOfItems = new int[noOfItems];
        
        System.out.println("\nEnter the corresponding weights for each item:");
        for (int i = 0; i < noOfItems; i++) {
            weightOfItems[i] = in.nextInt();
        }

        System.out.println("\nEnter the corresponding profits for each item:");
        for (int i = 0; i < noOfItems; i++) {
            profitOfItems[i] = in.nextInt();
        }

        System.out.println("\nThe entered data is as follows:\nItem\tWeight\tProfit");
        for (int i = 0; i < noOfItems; i++) {
            System.out.println(i + 1 + "\t" + weightOfItems[i] + "\t" + profitOfItems[i]);
        }

        int[][] knapSackArray = new int[noOfItems + 1][totalCapacity + 1];

        for (int i = 0; i < noOfItems; i++) {
            knapSackArray[i][0] = 0;
        }

        for (int j = 0; j < totalCapacity; j++) {
            knapSackArray[0][j] = 0;
        }

        System.out.println("\nSteps:");

        for (int i = 1; i <= noOfItems; i++) {
            for (int j = 1; j <= totalCapacity; j++) {
                if (weightOfItems[i - 1] <= j) {
                    knapSackArray[i][j] = Math.max(knapSackArray[i - 1][j],
                            knapSackArray[i - 1][j - weightOfItems[i - 1]] + profitOfItems[i - 1]);
                } else {
                    knapSackArray[i][j] = knapSackArray[i - 1][j];
                }
            }
        }

        display(knapSackArray, noOfItems, totalCapacity);
        
        System.out.println("\nMaximum profit is: " + knapSackArray[noOfItems][totalCapacity]);
        in.close();
    }
}

File: Dynamic Programming/knapSack.c
#include <stdio.h>
#include <stdlib.h>

int main() {
    int n, c;

    printf("\nEnter the number of items in the knapsack problem: ");
    scanf("%d", &n);
    printf("Enter the total capacity of the knapsack problem: ");
    scanf("%d", &c);

    int weightOfItems[n];
    int profitOfItems[n];

    printf("\nEnter the corresponding weights for each item:\n");
    for (int i = 0; i < n; i++) {
        scanf("%d", &weightOfItems[i]);
    }

    printf("\nEnter the corresponding profits for each item:\n");
    for (int i = 0; i < n; i++) {
        scanf("%d", &profitOfItems[i]);
    }

    printf("\nThe entered data is as follows:\nItem\tWeight\tProfit\n");
    for (int i = 0; i < n; i++) {
        printf("%d\t%d\t%d\n", i + 1, weightOfItems[i], profitOfItems[i]);
    }
    printf("\n");
    int knapSackArray[n][c];

    for (int i = 0; i < n; i++) {
        knapSackArray[i][0] = 0;
    }

    for (int j = 0; j < c; j++) {
        knapSackArray[0][j] = 0;
    }

    for (int i = 1; i < n; i++) {
        for (int j = 1; j < c; j++) {
            if (weightOfItems[i - 1] <= j) {
                knapSackArray[i][j] = (knapSackArray[i - 1][j] > 
                    (knapSackArray[i - 1][j - weightOfItems[i - 1]] + profitOfItems[i - 1])) ?
                    knapSackArray[i - 1][j] : 
                    (knapSackArray[i - 1][j - weightOfItems[i - 1]] + profitOfItems[i - 1]);
            } else {
                knapSackArray[i][j] = knapSackArray[i - 1][j];
            }
        }
    }

    printf("The knapsack table using dynamic programming is as follows:\n");
    for (int i = 0; i <= n; i++) {
        for (int j = 0; j <= c; j++) {
            printf("%d\t", knapSackArray[i][j]);
        }
        printf("\n");
    }

    printf("\nMaximum profit is: %d\n", knapSackArray[n][c]);

    return 0;
}